File: app/0003_Longest-Substring-Without-Repeating-Characters/main.cpp
#pragma once

#include <cassert>
#include <string>
#include <unordered_map>

namespace leetcode {

/// 依據 LeetCode 第 3 題設計，計算輸入字串中不含重複字元之最長子字串長度。
class LongestSubstringFinder final {
public:
    /// exec 為虛擬碼邏輯之執行進入點，回傳不含重複字元之最長子字串長度。
    [[nodiscard]] int exec(const std::string& inputString) const;
};

int LongestSubstringFinder::exec(const std::string& inputString) const {
    // 約束：輸入字串長度須介於 0 至 5 * 10^4 之間
    assert(inputString.size() <= 50000 && "輸入字串長度超出約束範圍");

    // 虛擬碼：若輸入字串為空，回傳 0
    if (inputString.empty()) {
        return 0;
    }

    // 虛擬碼：建立空對映，記錄字元對應最近出現之索引
    std::unordered_map<char, int> characterToLastIndex;
    int longestLength = 0;
    int windowStartIndex = 0;

    // 虛擬碼：走訪輸入字串，取得目前索引與目前字元
    for (int currentIndex = 0; currentIndex < static_cast<int>(inputString.size()); ++currentIndex) {
        const char currentCharacter = inputString[currentIndex];

        // 虛擬碼：若目前字元已存在於表中，且其索引落於目前視窗範圍內，則收縮視窗起始索引
        const auto foundIterator = characterToLastIndex.find(currentCharacter);
        if (foundIterator != characterToLastIndex.end() && foundIterator->second >= windowStartIndex) {
            windowStartIndex = foundIterator->second + 1;
        }

        // 虛擬碼：將目前字元與目前索引的對應關係加入字元對應最近索引表
        characterToLastIndex[currentCharacter] = currentIndex;

        // 虛擬碼：計算目前視窗長度，並更新最長長度
        const int currentWindowLength = currentIndex - windowStartIndex + 1;
        if (currentWindowLength > longestLength) {
            longestLength = currentWindowLength;
        }
    }

    // 虛擬碼：回傳最長長度
    return longestLength;
}

}  // namespace leetcode

File: app/0003_Longest-Substring-Without-Repeating-Characters/main.rb
# Longest Substring Without Repeating Characters
#
# 摘要：
# 給定一個字串 s，找出其中不含重複字元的最長子字串長度。
#
# 約束：
# - 0 <= s.length <= 5 * 10^4
# - s 由英文字母、數字、符號與空白組成。
#
# 註釋：
# - 回傳的是子字串的長度（整數），而非子字串本身。
# - 子字串（substring）須為連續字元序列，不可為子序列（subsequence）。
# - 當輸入字串為空時，回傳 0。
class LongestSubstringWithoutRepeatingCharacters
  # 執行虛擬碼邏輯：以滑動視窗（Sliding Window）搭配雜湊表記錄
  # 「字元 -> 最近索引」，動態調整視窗起始位置以避免重複字元。
  def exec(input_string)
    # 若 輸入字串 為空 則 回傳 0
    return 0 if input_string.empty?

    # 建立 空對映 字元對應最近索引表
    char_to_last_index = {}
    # 設 最長長度 為 0
    longest_length = 0
    # 設 視窗起始索引 為 0
    window_start_index = 0

    # 迴圈 走訪 輸入字串，取得 目前索引 與 目前字元
    input_string.each_char.with_index do |current_char, current_index|
      # 若 目前字元 存在於 字元對應最近索引表 中 且 對應之索引 大於等於 視窗起始索引 則
      if char_to_last_index.key?(current_char) && char_to_last_index[current_char] >= window_start_index
        # 設 視窗起始索引 為 字元對應最近索引表中目前字元對應之索引 加 1
        window_start_index = char_to_last_index[current_char] + 1
      end

      # 將 目前字元 與 目前索引 的對應關係 加入 字元對應最近索引表
      char_to_last_index[current_char] = current_index

      # 設 目前視窗長度 為 目前索引 減去 視窗起始索引 加 1
      current_window_length = current_index - window_start_index + 1
      # 若 目前視窗長度 大於 最長長度 則 設 最長長度 為 目前視窗長度
      longest_length = current_window_length if current_window_length > longest_length
    end

    # 回傳 最長長度
    longest_length
  end
end
